allo/ir/builder.py

```diff
@@ -58,16 +59,16 @@
     resolve_generic_types,
 )
 from .types import Int, UInt, Index, Float, Fixed, UFixed, Struct, float32
-from .visitor import ASTVisitor
+from .visitor import ASTVisitor, ASTContext, get_symbolic_expr
 from .symbol_resolver import ASTResolver
 from ..backend.ip import IPModule, c2allo_type
-from ..utils import get_mlir_dtype_from_str
+from ..utils import get_mlir_dtype_from_str, freeze_list, construct_kernel_name
 from ..logging import print_error_message
 from ..backend.aie.external_kernel import ExternalModule
 
 
 class ASTBuilder(ASTVisitor):
-    def __call__(self, ctx, node, file_name=None, **kwargs):
+    def __call__(self, ctx: ASTContext, node, file_name=None, **kwargs):
```

Comment:
If you were going to add type hints, could you also add for node, dtype, shape, etc.?